### PR TITLE
Make the Install-VSCode script work on *nix

### PR DESCRIPTION
**Pull request by rjmholt**
## PR Summary

Made a few changes so the `Install-VSCode.ps1` script supports macOS and various Linux distros as well.

It's probably not exhaustive and could do with a bit of testing.

I added a `-WhatIf` parameter as well, so you can use that to test and it will do the downloads but won't install anything.

## PR Checklist

Note: Tick the boxes below that apply to this pull request by putting an `x` between the square brackets.
Please mark anything not applicable to this PR `NA`.

- [x] PR has a meaningful title
- [x] Summarized changes
- [ ] PR has tests
- [x] This PR is ready to merge and is not work in progress
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready

### CODE REVIEW

File: scripts/Install-VSCode.ps1 [hunk 1]
```diff
@@ -133,118 +133,402 @@ param(
     [ValidateSet("Stable", "Insider-System", "Insider-User")]
     [string]$BuildEdition = "Stable",
 
-    [Parameter()]
-    [switch]$User,
```
Comment (rjmholt) [inline scripts/Install-VSCode.ps1]: This parameter didn't seem to be used...

Comment (tylerl0706) [reply scripts/Install-VSCode.ps1]: Hmm sure it wasn't used for User vs System install?

Comment (rjmholt) [reply scripts/Install-VSCode.ps1]: Yeah I had a search through. `$BuildEdition` already encodes that information, and it looks like `$User` meant to do that but got superseded and left in

File: scripts/Install-VSCode.ps1 [hunk 2]
```diff
@@ -133,118 +133,400 @@ param(
...
-        "64-bit" {
-            if ((Get-CimInstance -ClassName Win32_OperatingSystem).OSArchitecture -eq "64-bit") {
-                $codePath = $env:ProgramFiles
-                $bitVersion = "win32-x64"
+function Test-IsOsX64 {
+    if ($PSVersionTable.PSVersion.Major -lt 6) {
+        return (Get-CimInstance -ClassName Win32_OperatingSystem).OSArchitecture -eq "64-bit"
+    }
+
+    return [System.Runtime.InteropServices.RuntimeInformation]::OSArchitecture -eq [System.Runtime.InteropServices.Architecture]::X64
```
Comment (tylerl0706) [inline scripts/Install-VSCode.ps1]: You could probably use `using namespace System.Runtime.InteropServices`

File: scripts/Install-VSCode.ps1 [hunk 3]
```diff
@@ -133,118 +133,400 @@ param(
...
+
+function Get-CodePlatformInformation {
+    param(
+        [Parameter(Mandatory=$true)]
+        [ValidateSet('32-bit', '64-bit')]
+        [string]
+        $Bitness,
+
+        [Parameter(Mandatory=$true)]
+        [ValidateSet('Stable', 'Insider-System', 'Insider-User')]
```
[the remark this answers is not in the document]
Comment (tylerl0706) [reply scripts/Install-VSCode.ps1]: Consider having two parameters.

1. A switch to determine Insiders
2. A valid set (or another switch) between User and System installs

[the remark this answers is not in the document]
Comment (tylerl0706) [reply scripts/Install-VSCode.ps1]: Since this is an internal function, I don't care as much

File: scripts/Install-VSCode.ps1 [hunk 4]
```diff
@@ -133,118 +133,400 @@ param(
...
+            break
+        }
+    }
+
+    switch ($os) {
+        'Linux' {
+            $releaseInfo = Get-LinuxReleaseInfo
+
+            switch ($releaseInfo.NAME) {
+                { 'Ubuntu','Debian' -contains $_ } {
```
Comment (tylerl0706) [inline scripts/Install-VSCode.ps1]: Rip elementary lol If we can confirm that it works on elementary, I'll send a PR... I guess you cover general Linux below

File: scripts/Install-VSCode.ps1 [hunk 5]
```diff
@@ -133,118 +133,400 @@ param(
...
+    [switch]$WhatIf
 )
 
-if (($PSVersionTable.PSVersion.Major -le 5) -or $IsWindows) {
-    switch ($Architecture) {
-        "64-bit" {
-            if ((Get-CimInstance -ClassName Win32_OperatingSystem).OSArchitecture -eq "64-bit") {
-                $codePath = $env:ProgramFiles
-                $bitVersion = "win32-x64"
+function Test-IsOsX64 {
```
[the remark this answers is not in the document]
Comment (rjmholt) [reply scripts/Install-VSCode.ps1]: Have to distinguish from the process being 32/64-bit though

File: scripts/Install-VSCode.ps1 (hunk 3 above)
[the remark this answers is not in the document]
Comment (rjmholt) [reply scripts/Install-VSCode.ps1]: There's no stable user yet. This scheme matches the script parameters and prevents using `-User` with `Stable` for now. I also prefer having two switches, which will make sense later on. But I figure that will make sense to change when a stable user build comes out.

File: scripts/Install-VSCode.ps1 (hunk 2 above)
Comment (rjmholt) [reply scripts/Install-VSCode.ps1]: Is `using namespace` available in PowerShell v3?

File: scripts/Install-VSCode.ps1 (hunk 5 above)
[the remark this answers is not in the document]
Comment (rjmholt) [reply scripts/Install-VSCode.ps1]: And this only checks x86_64/amd64. If something returned arm64 (which `[system.runtime.interopservices.runtimeinformation]::OSArchitecture` can) this function would still return false

File: scripts/Install-VSCode.ps1 (hunk 4 above)
Comment (rjmholt) [reply scripts/Install-VSCode.ps1]: I actually thought about just testing if `apt`/`dpkg` vs `rpm` is available... Maybe we should do that instead?

File: scripts/Install-VSCode.ps1 (hunk 2 above)
Comment (tylerl0706) [reply scripts/Install-VSCode.ps1]: ugh... that's right.

File: scripts/Install-VSCode.ps1 [hunk 6]
```diff
@@ -322,14 +341,20 @@ function Get-CodePlatformInformation {
...
+    $info = @{
         AppName = $appName
         ExePath = $exePath
         Platform = $platform
         Channel = $channel
         FileUri = "https://vscode-update.azurewebsites.net/latest/$platform/$channel"
         Extension = $ext
     }
+
+    if ($pacMan) {
```
Comment (tylerl0706) [inline scripts/Install-VSCode.ps1]: You could probably just add this to the hashtable no?

File: scripts/Install-VSCode.ps1 [hunk 7]
```diff
@@ -133,118 +133,476 @@ param(
...
     [Parameter()]
     [ValidateNotNull()]
     [string[]]$AdditionalExtensions = @(),
 
     [switch]$LaunchWhenDone,
 
-    [switch]$EnableContextMenus
+    [switch]$EnableContextMenus,
+
+    [switch]$WhatIf
```
[the remark this answers is not in the document]
Comment (rjmholt) [reply scripts/Install-VSCode.ps1]: Ah I wondered about the right way for this -- I didn't find anything that discussed `WhatIf` very well. I'll change it :slightly_smiling_face:

File: scripts/Install-VSCode.ps1 [hunk 8]
```diff
@@ -462,7 +460,7 @@ try {
 
     $installerPath = [System.IO.Path]::Combine($tmpdir, $installerName)
 
-    if ($onWindows) {
+    if ($PSVersionTable.PSVersion.Major -lt 6) {
```
Comment (tylerl0706) [inline scripts/Install-VSCode.ps1]: Not Core compatible huh?

File: scripts/Install-VSCode.ps1 (hunk 7 above)
[the remark this answers is not in the document]
Comment (rjmholt) [reply scripts/Install-VSCode.ps1]: Ok @rkeithhill, I've implemented this with ShouldProcess now. Let me know if I'm using it the right way :)

File: scripts/Install-VSCode.ps1 (hunk 8 above)
Comment (rjmholt) [reply scripts/Install-VSCode.ps1]: BitsTransfer wasn't loading for me

File: scripts/Install-VSCode.ps1 [hunk 9]
```diff
@@ -133,118 +133,461 @@ param(
...
-    [switch]$User,
-
     [Parameter()]
     [ValidateNotNull()]
     [string[]]$AdditionalExtensions = @(),
 
     [switch]$LaunchWhenDone,
 
-    [switch]$EnableContextMenus
+    [switch]$EnableContextMenus,
```
Comment (SeeminglyScience) [inline scripts/Install-VSCode.ps1]: Wouldn't this <kbd>,</kbd> cause a parse exception?

```suggestion
    [switch]$EnableContextMenus
```

**Comment by rjmholt** [comment]
Ok, I've tested it on Windows and Ubuntu with WhatIf (and without -- it's part of an install script I wrote). Everything seems to be working :)

**Pull request merged**